Source/Csla.Data.EF7.Shared/DbContextManager.cs

```diff
@@ -0,0 +1,170 @@
+//-----------------------------------------------------------------------
+#if !MONO
+using System;
+using Microsoft.Data.Entity;
+
+namespace Csla.Data.EF7
+{
+  /// <summary>
+  /// Provides an automated way to reuse 
+  /// Entity Framework DbContext objects 
+  /// within the context
+  /// of a single data portal operation.
+  /// </summary>
+  /// <typeparam name="C">
+  /// Type of database 
+  /// DbContext object to use.
+  /// </typeparam>
+  /// <remarks>
+  /// This type stores the object context object 
+  /// in <see cref="Csla.ApplicationContext.LocalContext" />
+  /// and uses reference counting through
+  /// <see cref="IDisposable" /> to keep the data context object
+  /// open for reuse by child objects, and to automatically
+  /// dispose the object when the last consumer
+  /// has called Dispose."
+  /// </remarks>
+  public class DbContextManager<C> : IDisposable where C : DbContext
+  {
+    private static object _lock = new object();
+    private C _context;
+    private string _label;
+    private string ContextLabel { get; set; }
+
+    /// <summary>
+    /// Gets the ObjectContextManager object for the     /// specified database.
+    /// </summary>
+
+    public static DbContextManager<C> GetManager()
+    {
+      return GetManager(string.Empty);
+    }
+
+    /// <summary>
+    /// Gets the ObjectContextManager object for the 
+    /// specified database.
+    /// </summary>
+    /// <param name="database">Database name as shown in the config file.</param>
+    public static DbContextManager<C> GetManager(string database)
+    {
+      return GetManager(database, "default");
+    }
+   
+    /// <summary>
+    /// Gets the ObjectContextManager object for the 
+    /// specified database.
+    /// </summary>
+    /// <param name="database">
+    /// The database name or connection string.
+    /// </param>
+    /// <param name="label">Label for this context.</param>    
+    /// <returns>ContextManager object for the name.</returns>
+    public static DbContextManager<C> GetManager(string database, string label)
+    {
+      lock (_lock)
+      {
+        var contextLabel = GetContextName(database, label);
+        DbContextManager<C> mgr = null;
+        if (ApplicationContext.LocalContext.Contains(contextLabel))
+        {
+          mgr = (DbContextManager<C>)(ApplicationContext.LocalContext[contextLabel]);
+        }
+        else
+        {
+          mgr = new DbContextManager<C>(database, label);
+          mgr.ContextLabel = contextLabel;
+          ApplicationContext.LocalContext[contextLabel] = mgr;
+        }
+        mgr.AddRef();
+        return mgr;
+      }
+    }      
+
+    private DbContextManager(string database, string label)
+    {
+      _label = label;                
+      if (string.IsNullOrEmpty(database))
+      {
+        _context = (C) (Activator.CreateInstance(typeof (C)));
+      }
+      else
+      {
+        _context = (C)(Activator.CreateInstance(typeof(C), database));
```

Comment:
Have been thinking some more along the lines of what you suggested.
I'm wondering if we should take it further, and use dependency injection - via the new Microsoft.Extensions.DependencyInjection namespace (Previously Microsoft.Framework.DependencyInjection in beta releases)

This would work similar to what you are saying except on startup, you'd register your DbContext's (with their configured options) with an `IServiceCollection` and then `DbContextManager` would resolve the instance using `IServiceProvider`. Think this is going to complicate things somewhat, and perhaps venturing into the new DI land is not something I should attempt into this PR just yet..

```
DbContextManager<C, TOptionsFactory> GetManager(string label)

```